src/DynamoCore/Graph/Nodes/NodeModel.cs

```diff
@@ -1083,6 +1084,103 @@ protected NodeModel(IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPo
+                //TODO GET FEEDBACK - 
+                //a possibility is for variable port nodes to just give up here and do not
+                //modify the data coming from deserialization - risk is that these nodes may 
+                //create ports on the fly or do other transformations - just ask the node authors
+                //to handle this inside the json constructor - we've done the same for some variable input nodes.
+
+                // create a temporary node and extract data from it.
+                //TODO it would be great to avoid doing this for node types that we've already done it for...
+                //IE static or shared cache between node constructors - dumped/recreated after workspace load.
+
+                //TODO possible to create this node in another appdomain or somehow isolated?
+                var tempNode = this.GetType().GetConstructor(Type.EmptyTypes)?.Invoke(null) as NodeModel;
+                if(tempNode == null)
+                {
+                    //if the nodeModel did not have a parameterless constructor, let's give up!
+                    return;
+                }
+                inportDatas = tempNode.InPorts.Select(x => new PortData(x.Name, x.ToolTip)).ToList();
+                outportDatas = tempNode.outPorts.Select(x => new PortData(x.Name, x.ToolTip)).ToList();
+                tempNode.Dispose();
+
+                //if this was a variable input node we likey need to generate tooltips for the extra ports
+                //that users may have added.
+                var offset = inportDatas.Count();
+                if (this is VariableInputNode thisAsVarInput)
+                {
+                    for (int i = offset; i < inPorts.Count(); i++)
+                    {
+                        var data = thisAsVarInput.GetInputTooltipAndName(i);
+                        var newPortData = new PortData(data.name, data.tooltip);
+                        //add a new portdata for the existing port we deserialized.
+                        if (i >= inportDatas.Count())
+                        {
+                            inportDatas.Add(newPortData);
+                        }
+                        else
+                        {
+                            inportDatas[i] = newPortData;
+                        }
+                       
+                    }
+                }
+
+            }
+
+            for (var i = 0; i < outPorts.Count(); i++)
+            {
+                //bounds check
+                if (outportDatas.Count > i)
+                {
+                    var portData = outportDatas.ElementAt(i);
+                    var port = outPorts.ElementAt(i);
+                    port.Name = portData.Name == string.Empty ? port.Name : portData.Name;
+                    port.ToolTip = portData.ToolTipString == string.Empty ? port.ToolTip : portData.ToolTipString;
+                }
+            }
+
+            for (var i = 0; i < inPorts.Count(); i++)
```

Comment:
how about `for (var i = 0; i < inPorts.Count() && i < inportDatas.Count(); i++)`  without the bounds check ? ...same for `outPorts` above